README
This is the source code for a C++ debugger. It is designed to debug
applications (not kernel) written in C/C++ under Linux.

At its core is an engine that the user can interact with in a console,
somewhat like GDB. It also has a Gtk UI.

BUILDING THE SOURCE:
	Use the build script in this directory, type ./build --help for help
	DO NOT run "./configure; make" by hand!

BUILD ENVIRONMENT:
The following are required:
	g++
	libelf
	boost
	boost-python (dev)
	flex
	bison
	python 2.x (dev)
	libgtkmm-2.4 (dev)
	libgtksourceview-1.0 or 2.0 (dev)
	libgtksourceviewmm 2.0 (dev)
	libgtkhtml-3.x (dev)


SOURCE CODE LAYOUT:
engine: the debugger engine code

typez: an implementation of the DebugSymbol interface, used by the stabs and
        dwarf plug-ins, implementation of the type-system

dharma: from the Sanskrit word "that which supports": miscellaneous support
        code, such as C++ wrappers around libc and system calls, code for
        detecting and loading shared objects at runtime (plug-in support).

dwarfz: C++ wrapper library around libdwarf

elfz:   C++ wrapper library around libelf

interp: builtin C/C++ interpretor for one-liner expressions

zero_python: a library that exposes the ZDK to the Python programming language.
		Built on top of Boost.Python

zdk:    some abstract classes that serve as bases for most of the
        interfaces in the Zero debugger project: Unknown, RefCounted,
        EnumCallback<>;
        also contains some helper code such as the RefPtr<> smart pointer, 
        designed to work with RefCounted objects, and a RefCountedImpl<>
        template implementation;
        some support code for atomically incrementing/decrementing reference
        counters;
        and some architecture/platform-specific definitions 

zdk/generic: miscellaneous C++ templates, with generic support for the visitor
        pattern, smart ptrs, etc.

stabz:  library for reading STAB debug info

symbolz:library to read symbol tables, etc

plugin: contains the GUI plug-in, the DWARF and STABS readers, etc.

zui:	bootstraps an EXPERIMENTAL GUI written in Python, I use it every once in a blue moon
		for prototyping purposes.
		


ChangeLog.txt
I have stopped updating this file (and the commercial version).
For change history, please see:
http://zerobugs.codeplex.com/SourceControl/list/changesets.


[07/01/08]
- variables are now shown in tooltips when hovering over source code
- begun working on a utility for saving debug into to external database

[06/08/08]
- added missing headers (fixed compilation errors with g++ 4.3.0)
- added autoconfig test for unordered_map

[04/23/08]
- fixed bugs in RPC::RemoteIO, added test-client,
- moved object_factory.h from engine/ to dharma/

[01/23/08]
- Added license management, fixed bugs related to fork w/
multiple threads

[08/17/07]
- Added mutex to Dwarf::get_debug_handle, so only one thread
may operate on a Dwarf::Debug object.
- SymbolTable relatead tweaks, more experimentation with
caching

[07/21/07]
- Support for FPU regs in DWARF expressions on 32-bit systems
- UI fixes, gtk-- 1.2 related
- set "new_program" flag when thread is execv-ed
- made read_register non const, to allow for caching, etc.

[07/19/07]
- Fixed race conditions in symbol_table.cpp
- Fixed potential deadlock in UI
- Do not allow changing the view type for CodeView instances
that do now have a current scope function (symbol) set
- Added support in DebuggerShell for dumping symbol tables
- Revisited support for deferred breakpoints (i.e. insertion of
breakpoints in shared objects that have not yet been loaded) and
cleaned up the design; some small UI improvements entailed
TODO: persist deferred breakpoints
- Experimental feature (useful for own debugging): escape
temporarily from UI mode into text mode command line

[07/04/07]
Handle DW_FRAME_EXPRESSION, stack unwinding now works much better 
on AMD64

[06/27/07]
Do not install breakpoint for TD_DEATH, it seems to crash the main
thread, is this a known glibc bug?

And use SIGKILL rather than PTRACE_KILL to terminate the target,
PTRACE_KILL is flaky.

[06/23/07]
Migrated to SVN

[06/21/07]
Added support for reading symbols and debug info 
from separate file, when .gnu_debuglink is used

Remove threads from internal list upon PTRACE_EVENT_EXIT.

[06/19/07]
Fixed bug with displaying STL containers as arrays when aggregated
as struct/class members.

Added Watch window in bottom tab.
Added copyright note

[06/14/07]
Debugged core file are now properly added to the history.
Removed some compile-time dependencies in the UI.
Sorting in the SymbolTableImpl is now fully multi-threaded

[06/13/07]
Tinkered with ELF error reporting in minielf

[06/11/07]
Fixed unmangler assertion failure
Fixed bug in STAB parser: could not deal with nested structs/classes
Fixed problem with evaluating members of anonymous structs.
Fixed bug in DWARF reader wrt bit-fields
Experimenting with google::dense_hash_map, Hsieh function 

[06/07/07]
Fixed bugs in handling coredumps of 32-bit apps on AMD64
Fixed bugs in returning objects by value in the interpreter, on AMD64

Some small architectural cleanups: renamed the Task interface Runnable
and moved most methods that do not work on corefiles from Thread to Runnable

[05/06/07]
UI Fixes, assembly code is now tabbed nicely
Engine fix for returning from function when the PC is at func prologue
bumped revision to 54
 todo: test cases (would be interesting to play with recursive funcs)

[04/23/07]
Fixed install.py script
Fixed problem with the expression interpreter, which used to throw an
exception when attempting to pass an object by value, and the class did
not have a copy constructor defined.
A couple of things left to do in that area:
  Make it work with small (<= 16 bytes) objects on the x86-64, and
  use a member-wise copy rather than a bit-wise copy

[04/22/07]
Moved the pass_by_arg logic out of the interp library, to the target
directory (since it is architecture-specific)

[04/20/07]
Configure script exits if flex not found
Updated build instructions with libudis86 entry

[04/14/07]
Modified RPM spec file to invoke alien -k, if the alien program
is present

[04/12/07]
Fixed bug in eventlog (hex and dec modifiers applied to incorrect debug
levels)

Fixed the DynamicArray type (parse method used default implementation); to
prevent further errors, DataTypeImpl no longer provides a default
implementation for the parse method.

[04/07/07]
Fixed race condition in UI (related to right-menu)
Fixed DWARF-related BUG (module base was wrapping around)
FIXED UI bug (there was an attempt to update a stacktrace from the UI thread)
Support for D dynamic arrays
FIXED D demangling
New eventlog facility for the main engine and dwarf plugin (soon 
to be added to other modules)

DWARF plugin does not lookup all units in the module when looking
up line info by address (can be forced to do so with ZERO_AGGRESSIVE_LINEINFO,
just in case)

[03/39/07]

 1) fixed a bug in the parsing of STAB debug info (affected gcc 3.3.5 compiled
 code)
 2) fixed a bug in my implementation of ELF
 3) reduced the memory footprint for type information
 4) (internal) cleaned up my autoconfig scripts
 5) evaluation of logical expressions is now short-circuited (to better
 imitate the compiled-code); previously, both left-handside and righthandside
 operands were always being evaluated
 6) and fixed a bug that under certian conditions caused the debugger to hang
 after the debugged program exited successfully

 The stable version of Debian (Sarge) is now supported.


[12/24/06]
-- Fixed __x86_64__ / thread_db-related issue.

[08/24/06]
-- Allow static and implicit conversions from int and double
   to bool in the expression interpreter

[08/23/06]
-- Moved the TypeSystem instance to UnixTarget from under DebuggerEngine,
   this way when a process is forked the type system is not shared
-- Fixed a bug in step_until_current_func_returns dealing with recursive
   functions
-- Added  the E_THREAD_RETURNED event type to indicate that 
step_until_current_func_returns asynchronously completed

-- Fixed SEGV in zero_python error handling,
-- Fixed threading-related bug in the python plugin


plugin/flui/flview.h
#ifndef FLVIEW_H__2BBFE1FF_5D50_475B_87DD_3A119541D3BE
#define FLVIEW_H__2BBFE1FF_5D50_475B_87DD_3A119541D3BE
//
// -*- tab-width: 4; indent-tabs-mode: nil;  -*-
// vim: tabstop=4:softtabstop=4:expandtab:shiftwidth=4
//
// $Id: $
//
#include "view.h"

class Fl_Group;
class Fl_Widget;

void insert(Fl_Widget&, ui::LayoutCallback&);


template<typename W>
class FlViewImpl : public W
{
public:
    typedef W widget_type;

    template<typename... Args>
    FlViewImpl(Args... args) : W(args...)
    { }
};


struct FlViewBase
{
    virtual Fl_Widget* base_widget() = 0;
};


/**
 * Implements interface T in terms of a widget W.
 */
template<typename T, typename W>
class FlView : public T, public FlViewBase
{
public:
    template<typename... Args>
    FlView(ui::Controller& controller, Args... args)
        : T(controller)
        , impl_(new FlViewImpl<W>(args...))
    { }

    void resize(int x, int y, int w, int h) {
        base_widget()->resize(x, y, w, h);
    }

    W* widget() {
        assert(impl_);
        return impl_;
    }

    const W* widget() const {
        assert(impl_);
        return impl_;
    }

    Fl_Widget* base_widget() {
        return widget();
    }

protected:
    typedef FlView<T, W> base_type;

    virtual void insert_self(ui::LayoutCallback& cb) {
        insert(*impl_, cb);
    }
private:
    FlViewImpl<W>* impl_;
};

#endif // FLVIEW_H__2BBFE1FF_5D50_475B_87DD_3A119541D3BE



third_party_libs/dwarf-20111214/dwarfdump/print_reloc.h
/* 
  Copyright (C) 2000,2004,2005 Silicon Graphics, Inc.  All Rights Reserved.
  Portions Copyright (C) 2007-2011 David Anderson. All Rights Reserved.
  Portions Copyright (C) 2011 SN Systems Ltd. All rights reserved.

  This program is free software; you can redistribute it and/or modify it
  under the terms of version 2 of the GNU General Public License as
  published by the Free Software Foundation.

  This program is distributed in the hope that it would be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

  Further, this software is distributed without any warranty that it is
  free of the rightful claim of any third person regarding infringement
  or the like.  Any license provided herein, whether implied or
  otherwise, applies only to this software file.  Patent licenses, if
  any, provided herein do not apply to combinations of this program with
  other software, or any other product whatsoever.

  You should have received a copy of the GNU General Public License along
  with this program; if not, write the Free Software Foundation, Inc., 51
  Franklin Street - Fifth Floor, Boston MA 02110-1301, USA.

  Contact information:  Silicon Graphics, Inc., 1500 Crittenden Lane,
  Mountain View, CA 94043, or:

  http://www.sgi.com

  For further information regarding this notice, see:

  http://oss.sgi.com/projects/GenInfo/NoticeExplan

$Header: /plroot/cmplrs.src/v7.4.5m/.RCS/PL/dwarfdump/RCS/print_reloc.c,v 1.11 2005/08/04 05:09:37 davea Exp $ */

/* The address of the Free Software Foundation is
   Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, 
   Boston, MA 02110-1301, USA.
   SGI has moved from the Crittenden Lane address.
*/

/* PowerPC relocations defined by the ABIs */
#define R_PPC_NONE                   0
#define R_PPC_ADDR32                 1 /* 32bit absolute address */
#define R_PPC_ADDR24                 2 /* 26bit address, 2 bits ignored.  */
#define R_PPC_ADDR16                 3 /* 16bit absolute address */
#define R_PPC_ADDR16_LO              4 /* lower 16bit of absolute address */
#define R_PPC_ADDR16_HI              5 /* high 16bit of absolute address */
#define R_PPC_ADDR16_HA              6 /* adjusted high 16bit */
#define R_PPC_ADDR14                 7 /* 16bit address, 2 bits ignored */
#define R_PPC_ADDR14_BRTAKEN         8
#define R_PPC_ADDR14_BRNTAKEN        9
#define R_PPC_REL24                 10 /* PC relative 26 bit */
#define R_PPC_REL14                 11 /* PC relative 16 bit */
#define R_PPC_REL14_BRTAKEN         12
#define R_PPC_REL14_BRNTAKEN        13
#define R_PPC_GOT16                 14
#define R_PPC_GOT16_LO              15
#define R_PPC_GOT16_HI              16
#define R_PPC_GOT16_HA              17
#define R_PPC_PLTREL24              18
#define R_PPC_COPY                  19
#define R_PPC_GLOB_DAT              20
#define R_PPC_JMP_SLOT              21
#define R_PPC_RELATIVE              22
#define R_PPC_LOCAL24PC             23
#define R_PPC_UADDR32               24
#define R_PPC_UADDR16               25
#define R_PPC_REL32                 26
#define R_PPC_PLT32                 27
#define R_PPC_PLTREL32              28
#define R_PPC_PLT16_LO              29
#define R_PPC_PLT16_HI              30
#define R_PPC_PLT16_HA              31
#define R_PPC_SDAREL16              32
#define R_PPC_SECTOFF               33
#define R_PPC_SECTOFF_LO            34
#define R_PPC_SECTOFF_HI            35
#define R_PPC_SECTOFF_HA            36

/* Unused types */
#define R_PPC_37                    37
#define R_PPC_38                    38
#define R_PPC_39                    39
#define R_PPC_40                    40
#define R_PPC_41                    41
#define R_PPC_42                    42
#define R_PPC_43                    43
#define R_PPC_44                    44
#define R_PPC_45                    45
#define R_PPC_46                    46
#define R_PPC_47                    47
#define R_PPC_48                    48
#define R_PPC_49                    49
#define R_PPC_50                    50
#define R_PPC_51                    51
#define R_PPC_52                    52
#define R_PPC_53                    53
#define R_PPC_54                    54
#define R_PPC_55                    55

/* Unused types */
#define R_PPC_56                    56
#define R_PPC_57                    57
#define R_PPC_58                    58
#define R_PPC_59                    59
#define R_PPC_60                    60
#define R_PPC_61                    61
#define R_PPC_62                    62
#define R_PPC_63                    63
#define R_PPC_64                    64
#define R_PPC_65                    65
#define R_PPC_66                    66

/* PowerPC relocations defined for the TLS access ABI.  */
#define R_PPC_TLS                   67 /* none      (sym+add)@tls */
#define R_PPC_DTPMOD32              68 /* word32    (sym+add)@dtpmod */
#define R_PPC_TPREL16               69 /* half16*   (sym+add)@tprel */
#define R_PPC_TPREL16_LO            70 /* half16    (sym+add)@tprel@l */
#define R_PPC_TPREL16_HI            71 /* half16    (sym+add)@tprel@h */
#define R_PPC_TPREL16_HA            72 /* half16    (sym+add)@tprel@ha */
#define R_PPC_TPREL32               73 /* word32    (sym+add)@tprel */
#define R_PPC_DTPREL16              74 /* half16*   (sym+add)@dtprel */
#define R_PPC_DTPREL16_LO           75 /* half16    (sym+add)@dtprel@l */
#define R_PPC_DTPREL16_HI           76 /* half16    (sym+add)@dtprel@h */
#define R_PPC_DTPREL16_HA           77 /* half16    (sym+add)@dtprel@ha */
#define R_PPC_DTPREL32              78 /* word32    (sym+add)@dtprel */
#define R_PPC_GOT_TLSGD16           79 /* half16*   (sym+add)@got@tlsgd */
#define R_PPC_GOT_TLSGD16_LO        80 /* half16    (sym+add)@got@tlsgd@l */
#define R_PPC_GOT_TLSGD16_HI        81 /* half16    (sym+add)@got@tlsgd@h */
#define R_PPC_GOT_TLSGD16_HA        82 /* half16    (sym+add)@got@tlsgd@ha */
#define R_PPC_GOT_TLSLD16           83 /* half16*   (sym+add)@got@tlsld */
#define R_PPC_GOT_TLSLD16_LO        84 /* half16    (sym+add)@got@tlsld@l */
#define R_PPC_GOT_TLSLD16_HI        85 /* half16    (sym+add)@got@tlsld@h */
#define R_PPC_GOT_TLSLD16_HA        86 /* half16    (sym+add)@got@tlsld@ha */
#define R_PPC_GOT_TPREL16           87 /* half16*   (sym+add)@got@tprel */
#define R_PPC_GOT_TPREL16_LO        88 /* half16    (sym+add)@got@tprel@l */
#define R_PPC_GOT_TPREL16_HI        89 /* half16    (sym+add)@got@tprel@h */
#define R_PPC_GOT_TPREL16_HA        90 /* half16    (sym+add)@got@tprel@ha */
#define R_PPC_GOT_DTPREL16          91 /* half16*   (sym+add)@got@dtprel */
#define R_PPC_GOT_DTPREL16_LO       92 /* half16*   (sym+add)@got@dtprel@l */
#define R_PPC_GOT_DTPREL16_HI       93 /* half16*   (sym+add)@got@dtprel@h */
#define R_PPC_GOT_DTPREL16_HA       94 /* half16*   (sym+add)@got@dtprel@ha */

/* Keep this the last entry.  */
#define R_PPC_NUM                   95

/*  PowerPC64 relocations defined by the ABIs */
#define R_PPC64_NONE                R_PPC_NONE
#define R_PPC64_ADDR32              R_PPC_ADDR32  /* 32bit absolute address.  */
#define R_PPC64_ADDR24              R_PPC_ADDR24  /* 26bit address, word aligned.  */
#define R_PPC64_ADDR16              R_PPC_ADDR16  /* 16bit absolute address. */
#define R_PPC64_ADDR16_LO           R_PPC_ADDR16_LO /* lower 16bits of abs. address.  */
#define R_PPC64_ADDR16_HI           R_PPC_ADDR16_HI /* high 16bits of abs. address. */
#define R_PPC64_ADDR16_HA           R_PPC_ADDR16_HA /* adjusted high 16bits.  */
#define R_PPC64_ADDR14              R_PPC_ADDR14   /* 16bit address, word aligned.  */
#define R_PPC64_ADDR14_BRTAKEN      R_PPC_ADDR14_BRTAKEN
#define R_PPC64_ADDR14_BRNTAKEN     R_PPC_ADDR14_BRNTAKEN
#define R_PPC64_REL24               R_PPC_REL24 /* PC relative 26 bit, word aligned.  */
#define R_PPC64_REL14               R_PPC_REL14 /* PC relative 16 bit. */
#define R_PPC64_REL14_BRTAKEN       R_PPC_REL14_BRTAKEN
#define R_PPC64_REL14_BRNTAKEN      R_PPC_REL14_BRNTAKEN
#define R_PPC64_GOT16               R_PPC_GOT16
#define R_PPC64_GOT16_LO            R_PPC_GOT16_LO
#define R_PPC64_GOT16_HI            R_PPC_GOT16_HI
#define R_PPC64_GOT16_HA            R_PPC_GOT16_HA

#define R_PPC64_COPY                R_PPC_COPY
#define R_PPC64_GLOB_DAT            R_PPC_GLOB_DAT
#define R_PPC64_JMP_SLOT            R_PPC_JMP_SLOT
#define R_PPC64_RELATIVE            R_PPC_RELATIVE

#define R_PPC64_UADDR32             R_PPC_UADDR32
#define R_PPC64_UADDR16             R_PPC_UADDR16
#define R_PPC64_REL32               R_PPC_REL32
#define R_PPC64_PLT32               R_PPC_PLT32
#define R_PPC64_PLTREL32            R_PPC_PLTREL32
#define R_PPC64_PLT16_LO            R_PPC_PLT16_LO
#define R_PPC64_PLT16_HI            R_PPC_PLT16_HI
#define R_PPC64_PLT16_HA            R_PPC_PLT16_HA

#define R_PPC64_SECTOFF             R_PPC_SECTOFF
#define R_PPC64_SECTOFF_LO          R_PPC_SECTOFF_LO
#define R_PPC64_SECTOFF_HI          R_PPC_SECTOFF_HI
#define R_PPC64_SECTOFF_HA          R_PPC_SECTOFF_HA
#define R_PPC64_ADDR30              37  /* word30 (S + A - P) >> 2.  */
#define R_PPC64_ADDR64              38  /* doubleword64 S + A.  */
#define R_PPC64_ADDR16_HIGHER       39  /* half16 #higher(S + A).  */
#define R_PPC64_ADDR16_HIGHERA      40  /* half16 #highera(S + A).  */
#define R_PPC64_ADDR16_HIGHEST      41  /* half16 #highest(S + A).  */
#define R_PPC64_ADDR16_HIGHESTA     42  /* half16 #highesta(S + A). */
#define R_PPC64_UADDR64             43  /* doubleword64 S + A.  */
#define R_PPC64_REL64               44  /* doubleword64 S + A - P.  */
#define R_PPC64_PLT64               45  /* doubleword64 L + A.  */
#define R_PPC64_PLTREL64            46  /* doubleword64 L + A - P.  */
#define R_PPC64_TOC16               47  /* half16* S + A - .TOC.  */
#define R_PPC64_TOC16_LO            48  /* half16 #lo(S + A - .TOC.).  */
#define R_PPC64_TOC16_HI            49  /* half16 #hi(S + A - .TOC.).  */
#define R_PPC64_TOC16_HA            50  /* half16 #ha(S + A - .TOC.).  */
#define R_PPC64_TOC                 51  /* doubleword64 .TOC. */
#define R_PPC64_PLTGOT16            52  /* half16* M + A.  */
#define R_PPC64_PLTGOT16_LO         53  /* half16 #lo(M + A).  */
#define R_PPC64_PLTGOT16_HI         54  /* half16 #hi(M + A).  */
#define R_PPC64_PLTGOT16_HA         55  /* half16 #ha(M + A).  */

#define R_PPC64_ADDR16_DS           56 /* half16ds* (S + A) >> 2.  */
#define R_PPC64_ADDR16_LO_DS        57 /* half16ds  #lo(S + A) >> 2.  */
#define R_PPC64_GOT16_DS            58 /* half16ds* (G + A) >> 2.  */
#define R_PPC64_GOT16_LO_DS         59 /* half16ds  #lo(G + A) >> 2.  */
#define R_PPC64_PLT16_LO_DS         60 /* half16ds  #lo(L + A) >> 2.  */
#define R_PPC64_SECTOFF_DS          61 /* half16ds* (R + A) >> 2.  */
#define R_PPC64_SECTOFF_LO_DS       62 /* half16ds  #lo(R + A) >> 2.  */
#define R_PPC64_TOC16_DS            63 /* half16ds* (S + A - .TOC.) >> 2.  */
#define R_PPC64_TOC16_LO_DS         64 /* half16ds  #lo(S + A - .TOC.) >> 2.  */
#define R_PPC64_PLTGOT16_DS         65 /* half16ds* (M + A) >> 2.  */
#define R_PPC64_PLTGOT16_LO_DS      66 /* half16ds  #lo(M + A) >> 2.  */

/* PowerPC64 relocations defined for the TLS access ABI.  */
#define R_PPC64_TLS                 67 /* none      (sym+add)@tls */
#define R_PPC64_DTPMOD64            68 /* doubleword64 (sym+add)@dtpmod */
#define R_PPC64_TPREL16             69 /* half16*   (sym+add)@tprel */
#define R_PPC64_TPREL16_LO          70 /* half16    (sym+add)@tprel@l */
#define R_PPC64_TPREL16_HI          71 /* half16    (sym+add)@tprel@h */
#define R_PPC64_TPREL16_HA          72 /* half16    (sym+add)@tprel@ha */
#define R_PPC64_TPREL64             73 /* doubleword64 (sym+add)@tprel */
#define R_PPC64_DTPREL16            74 /* half16*   (sym+add)@dtprel */
#define R_PPC64_DTPREL16_LO         75 /* half16    (sym+add)@dtprel@l */
#define R_PPC64_DTPREL16_HI         76 /* half16    (sym+add)@dtprel@h */
#define R_PPC64_DTPREL16_HA         77 /* half16    (sym+add)@dtprel@ha */
#define R_PPC64_DTPREL64            78 /* doubleword64 (sym+add)@dtprel */
#define R_PPC64_GOT_TLSGD16         79 /* half16*   (sym+add)@got@tlsgd */
#define R_PPC64_GOT_TLSGD16_LO      80 /* half16    (sym+add)@got@tlsgd@l */
#define R_PPC64_GOT_TLSGD16_HI      81 /* half16    (sym+add)@got@tlsgd@h */
#define R_PPC64_GOT_TLSGD16_HA      82 /* half16    (sym+add)@got@tlsgd@ha */
#define R_PPC64_GOT_TLSLD16         83 /* half16*   (sym+add)@got@tlsld */
#define R_PPC64_GOT_TLSLD16_LO      84 /* half16    (sym+add)@got@tlsld@l */
#define R_PPC64_GOT_TLSLD16_HI      85 /* half16    (sym+add)@got@tlsld@h */
#define R_PPC64_GOT_TLSLD16_HA      86 /* half16    (sym+add)@got@tlsld@ha */
#define R_PPC64_GOT_TPREL16_DS      87 /* half16ds* (sym+add)@got@tprel */
#define R_PPC64_GOT_TPREL16_LO_DS   88 /* half16ds (sym+add)@got@tprel@l */
#define R_PPC64_GOT_TPREL16_HI      89 /* half16    (sym+add)@got@tprel@h */
#define R_PPC64_GOT_TPREL16_HA      90 /* half16    (sym+add)@got@tprel@ha */
#define R_PPC64_GOT_DTPREL16_DS     91 /* half16ds* (sym+add)@got@dtprel */
#define R_PPC64_GOT_DTPREL16_LO_DS  92 /* half16ds (sym+add)@got@dtprel@l */
#define R_PPC64_GOT_DTPREL16_HI     93 /* half16    (sym+add)@got@dtprel@h */
#define R_PPC64_GOT_DTPREL16_HA     94 /* half16    (sym+add)@got@dtprel@ha */
#define R_PPC64_TPREL16_DS          95 /* half16ds* (sym+add)@tprel */
#define R_PPC64_TPREL16_LO_DS       96 /* half16ds  (sym+add)@tprel@l */
#define R_PPC64_TPREL16_HIGHER      97 /* half16    (sym+add)@tprel@higher */
#define R_PPC64_TPREL16_HIGHERA     98 /* half16    (sym+add)@tprel@highera */
#define R_PPC64_TPREL16_HIGHEST     99 /* half16    (sym+add)@tprel@highest */
#define R_PPC64_TPREL16_HIGHESTA   100 /* half16  (sym+add)@tprel@highesta */
#define R_PPC64_DTPREL16_DS        101 /* half16ds* (sym+add)@dtprel */
#define R_PPC64_DTPREL16_LO_DS     102 /* half16ds (sym+add)@dtprel@l */
#define R_PPC64_DTPREL16_HIGHER    103 /* half16   (sym+add)@dtprel@higher */
#define R_PPC64_DTPREL16_HIGHERA   104 /* half16  (sym+add)@dtprel@highera */
#define R_PPC64_DTPREL16_HIGHEST   105 /* half16  (sym+add)@dtprel@highest */
#define R_PPC64_DTPREL16_HIGHESTA  106 /* half16 (sym+add)@dtprel@highesta */

/*  Additional relocation types */
#define R_PPC64_TOC32              107
#define R_PPC64_DTPMOD32           108
#define R_PPC64_TPREL32            109
#define R_PPC64_DTPREL32           110

/*  Keep this the last entry.  */
#define R_PPC64_NUM                111

/*  Relocation types for MIPS */
#define R_MIPS_NONE                   0
#define R_MIPS_16                     1
#define R_MIPS_32                     2
#define R_MIPS_ADD                    2
#define R_MIPS_REL                    3
#define R_MIPS_REL32                  3
#define R_MIPS_26                     4
#define R_MIPS_HI16                   5
#define R_MIPS_LO16                   6
#define R_MIPS_GPREL                  7
#define R_MIPS_GPREL16                7
#define R_MIPS_LITERAL                8
#define R_MIPS_GOT                    9
#define R_MIPS_GOT16                  9
#define R_MIPS_PC16                  10
#define R_MIPS_CALL                  11
#define R_MIPS_CALL16                11
#define R_MIPS_GPREL32               12

#define R_MIPS_SHIFT5                16
#define R_MIPS_SHIFT6                17
#define R_MIPS_64                    18
#define R_MIPS_GOT_DISP              19
#define R_MIPS_GOT_PAGE              20
#define R_MIPS_GOT_OFST              21
#define R_MIPS_GOT_HI16              22
#define R_MIPS_GOT_LO16              23
#define R_MIPS_SUB                   24
#define R_MIPS_INSERT_A              25
#define R_MIPS_INSERT_B              26
#define R_MIPS_DELETE                27
#define R_MIPS_HIGHER                28
#define R_MIPS_HIGHEST               29
#define R_MIPS_CALL_HI16             30
#define R_MIPS_CALL_LO16             31
#define R_MIPS_SCN_DISP              32
#define	R_MIPS_REL16                 33
#define R_MIPS_ADD_IMMEDIATE         34
